cpp/tests/utilities/column_utilities.cu

```diff
@@ -536,28 +540,47 @@ struct column_comparator_impl {
                   size_type fp_ulps,
                   int depth)
   {
-    auto d_lhs = cudf::table_device_view::create(table_view{{lhs}});
-    auto d_rhs = cudf::table_device_view::create(table_view{{rhs}});
+    auto lhs_tview = table_view{{lhs}};
+    auto rhs_tview = table_view{{rhs}};
+
+    auto d_lhs = cudf::table_device_view::create(lhs_tview);
+    auto d_rhs = cudf::table_device_view::create(rhs_tview);
 
     auto d_lhs_row_indices = cudf::column_device_view::create(lhs_row_indices);
     auto d_rhs_row_indices = cudf::column_device_view::create(rhs_row_indices);
 
-    using ComparatorType = std::conditional_t<check_exact_equality,
-                                              corresponding_rows_unequal,
-                                              corresponding_rows_not_equivalent>;
-
     auto differences = rmm::device_uvector<int>(
       lhs.size(), cudf::get_default_stream());  // worst case: everything different
     auto input_iter = thrust::make_counting_iterator(0);
-    auto diff_iter  = thrust::copy_if(
-      rmm::exec_policy(cudf::get_default_stream()),
-      input_iter,
-      input_iter + lhs_row_indices.size(),
-      differences.begin(),
-      ComparatorType(*d_lhs, *d_rhs, *d_lhs_row_indices, *d_rhs_row_indices, fp_ulps));
 
-    differences.resize(thrust::distance(differences.begin(), diff_iter),
-                       cudf::get_default_stream());  // shrink back down
+    auto const comparator_helper = [&](auto const device_comparator) {
+      using ComparatorType =
+        std::conditional_t<check_exact_equality,
+                           corresponding_rows_unequal<decltype(device_comparator)>,
+                           corresponding_rows_not_equivalent<decltype(device_comparator)>>;
+
+      auto diff_iter = thrust::copy_if(
+        rmm::exec_policy(cudf::get_default_stream()),
+        input_iter,
+        input_iter + lhs_row_indices.size(),
+        differences.begin(),
+        ComparatorType(
+          *d_lhs, *d_rhs, *d_lhs_row_indices, *d_rhs_row_indices, fp_ulps, device_comparator));
+
+      differences.resize(thrust::distance(differences.begin(), diff_iter),
+                         cudf::get_default_stream());  // shrink back down
+    };
+
+    auto const comparator = cudf::experimental::row::equality::two_table_comparator{
+      lhs_tview, rhs_tview, cudf::get_default_stream()};
+    auto const has_nulls = cudf::has_nested_nulls(lhs_tview) or cudf::has_nested_nulls(rhs_tview);
+    if (cudf::detail::has_nested_columns(lhs_tview)) {
```

Comment:
This struct is specialized to work with non-nested types. So this should not be any `if/else`.